### PR TITLE
Release 0.1 - Device Authorizer with persistent backend

### PR DESCRIPTION
**Pull request by nivemaham**
An application to get authorization from device users to access their data through 3rd party APIs. Mainly supports OAuth2 Authorization code follow to request authorization and processes the returned authorization code to add new entries of users. It creates new users and adds properties compatible and required by the `RADAR-REST-CONNECTOR/RADAR-FITBIT-REST-CONNECTOR`.
- It has one active entity where we store user properties.
- Has `liquibase` support to enable seamless database schema migration.
- Has a simple web-service with REST Endpoints to share configured device-type details and authorized users.
- Currently various device-types can be configured using a YAML file and these entries are stored in memory.
- Has a simple UI to interact with user to enable device authorization.

### CODE REVIEW

File: authorizer-app/src/environments/environment.prod.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
 export const environment = {
   production: true,
-  BACKEND_BASE_URL: '/device-auth/backend'
+  BACKEND_BASE_URL: '/rest-sources/backend'
```
Comment (yatharthranjan) [inline authorizer-app/src/environments/environment.prod.ts]: is this the REST endpoint of the api ?? If yes, perhaps `/rest-sources/api/` ?

Comment (nivemaham) [reply authorizer-app/src/environments/environment.prod.ts]: this is the route served by nginx at the end. For production this paths need to be mentioned before compile time.

Comment (yatharthranjan) [reply authorizer-app/src/environments/environment.prod.ts]: Ok, Thanks.

File: src/main/java/org/radarbase/authorizer/webapp/resource/DeviceClientResource.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+@RestController
+public class DeviceClientResource {
+
+    private Logger logger = LoggerFactory.getLogger(DeviceClientResource.class);
+
+    @Autowired
+    private DeviceClientService deviceClientService;
+
+    @GetMapping("/device-clients")
```
[the remark this answers is not in the document]
Comment (nivemaham) [reply src/main/java/org/radarbase/authorizer/webapp/resource/DeviceClientResource.java]: This refers to the oauth-clients of the source. I will change it to source-clients

File: authorizer-app/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,23 @@
+FROM node:8.12.0-alpine as builder
+
+WORKDIR /app
+
+COPY package.json /app/
```
Comment (blootsvoets) [inline authorizer-app/Dockerfile]: also copy package-lock.json

File: authorizer-app/.gitignore [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+npm-debug.log
+yarn-error.log
+testem.log
+/typings
+
+# System Files
+.DS_Store
+Thumbs.db
+
+package-lock.json
```
Comment (blootsvoets) [inline authorizer-app/.gitignore]: The package lock should not be ignored.

File: README.md [hunk 5]
```diff
@@ -1,11 +1,32 @@
-# RADAR-FitBit-Auth
+# RADAR-Device-Auth
```
Comment (blootsvoets) [inline README.md]:
```suggestion
# RADAR-REST-Source-Auth
```

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+/out/
+
+### NetBeans ###
+/nbproject/private/
+/nbbuild/
+/dist/
+/nbdist/
+/.nb-gradle/
+
+/src/main/docker/etc/device_clients_configs.yml
```
Comment (blootsvoets) [inline .gitignore]:
```suggestion
/src/main/docker/etc/radar_rest_source_auth_config.yml
```

File: src/main/docker/app.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+services:
+    radar-auth-backend:
+        image: radarbase/radar-device-auth-backend:latest
+        depends_on:
+          - auth-postgresql
+        environment:
+            - SPRING_DATASOURCE_URL=jdbc:postgresql://auth-postgresql:5432/deviceauthorizer
+            - SPRING_DATASOURCE_USERNAME=radarcns
+            - SPRING_DATASOURCE_PASSWORD=radarcns
+            - DEVICE_AUTHORIZER_DEVICE_CLIENTS_FILE_PATH=app-includes/device_clients_configs.yml
```
Comment (blootsvoets) [inline src/main/docker/app.yml]:
```suggestion
            - RADAR_REST_SOURCE_AUTHORIZER_CONFIG_PATH=app-includes/radar_rest_source_auth_config.yml
```

File: src/main/docker/app.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
+version: '2'
+services:
+    radar-auth-backend:
+        image: radarbase/radar-device-auth-backend:latest
+        depends_on:
+          - auth-postgresql
+        environment:
+            - SPRING_DATASOURCE_URL=jdbc:postgresql://auth-postgresql:5432/deviceauthorizer
```
Comment (blootsvoets) [inline src/main/docker/app.yml]:
```suggestion
            - SPRING_DATASOURCE_URL=jdbc:postgresql://auth-postgresql:5432/restsourceauth
```

File: src/main/docker/app.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        volumes:
+            - ./etc:/app-includes
+    auth-postgresql:
+        image: postgres:10.4-alpine
+#        ports:
+#          - "5432:5432"
+        environment:
+          - POSTGRES_USER=radarcns
+          - POSTGRES_PASSWORD=radarcns
+          - POSTGRES_DB=deviceauthorizer
```
Comment (blootsvoets) [inline src/main/docker/app.yml]:
```suggestion
          - POSTGRES_DB=restsourceauth
```

File: src/main/docker/app.yml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+    radar-auth-frontend:
+        image: radarbase/radar-device-authorizer:latest
```
Comment (blootsvoets) [inline src/main/docker/app.yml]:
```suggestion
        image: radarbase/radar-rest-source-authorizer:latest
```

File: src/main/docker/postgresql.yml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+version: '2'
+services:
+    postgresql:
+        image: postgres:10.4-alpine
+        ports:
+          - "5432:5432"
+        environment:
+          - POSTGRES_USER=radarcns
+          - POSTGRES_PASSWORD=radarcns
+          - POSTGRES_DB=deviceauthorizer
```
Comment (blootsvoets) [inline src/main/docker/postgresql.yml]:
```suggestion
          - POSTGRES_DB=restsourceauth
```

File: src/main/java/org/radarbase/authorizer/RadarDeviceAuthorizerApplication.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.web.cors.CorsConfiguration;
+import org.springframework.web.servlet.config.annotation.CorsRegistry;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
+
+@SpringBootApplication
+@EnableConfigurationProperties({DeviceAuthorizerApplicationProperties.class})
+public class RadarDeviceAuthorizerApplication {
```
Comment (blootsvoets) [inline src/main/java/org/radarbase/authorizer/RadarDeviceAuthorizerApplication.java]:
```suggestion
public class RadarRestSourceAuthorizerApplication {
```

File: src/main/docker/etc/device_clients_configs.yml.template [hunk 13]
```diff
@@ -0,0 +1,7 @@
+device_clients:
```
Comment (blootsvoets) [inline src/main/docker/etc/device_clients_configs.yml.template]: rest_source_clients

File: src/main/docker/etc/device_clients_configs.yml.template [hunk 14]
```diff
@@ -0,0 +1,7 @@
...
+  - device_type: FitBit
```
Comment (blootsvoets) [inline src/main/docker/etc/device_clients_configs.yml.template]:
```suggestion
  - source_type: FitBit
```

File: src/main/docker/etc/device_clients_configs.yml.template [hunk 15]
```diff
@@ -0,0 +1,7 @@
...
+    authorization_endpoint: https://www.fitbit.com/oauth2/authorize
+    token_endpoint: https://api.fitbit.com/oauth2/token
+    client_id:
```
Comment (blootsvoets) [inline src/main/docker/etc/device_clients_configs.yml.template]:
```suggestion
    client_id: MY_CLIENT_ID
```

File: src/main/docker/etc/device_clients_configs.yml.template [hunk 16]
```diff
@@ -0,0 +1,7 @@
+device_clients:
+  - device_type: FitBit
+    authorization_endpoint: https://www.fitbit.com/oauth2/authorize
+    token_endpoint: https://api.fitbit.com/oauth2/token
+    client_id: 
+    client_secret:
```
Comment (blootsvoets) [inline src/main/docker/etc/device_clients_configs.yml.template]:
```suggestion
    client_secret: MY_CLIENT_SECRET
```

File: README.md [hunk 17]
```diff
@@ -1,11 +1,32 @@
...
+2. Has liquibase support to enable seamless database schema migration.
+3. Has a simple web-service with REST Endpoints to share configured source-type client details and authorized users.
+4. Currently various source-types can be configured using a YAML file and these entries are stored in memory.
 
-    mvn clean install
-    cd fitbit/fitbit-rest-service
-    mvn clean install
-    cd ..
-    docker-compose up -d
+## APIs to be used by Device-Connectors
```
Comment (blootsvoets) [inline README.md]:
```suggestion
## APIs to be used by RADAR REST Connectors
```

File: README.md [hunk 18]
```diff
@@ -1,11 +1,32 @@
...
+Device `Source-Connectors` can use the APIs as follows.
```
Comment (blootsvoets) [inline README.md]:
```suggestion
RADAR REST Source Connectors can use the APIs as follows.
```

File: authorizer-app/src/app/models/device-auth-details.model.ts [hunk 19]
```diff
@@ -0,0 +1,8 @@
+export class DeviceAuthDetails {
+  deviceType?: string;
```
Comment (blootsvoets) [inline authorizer-app/src/app/models/device-auth-details.model.ts]:
```suggestion
  sourceType?: string;
```

File: build.gradle [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+	maven { url 'https://oss.jfrog.org/artifactory/libs-snapshot/' }
+	maven { url 'https://dl.bintray.com/radar-cns/org.radarcns' }
+}
+
+ext {
+	githubRepoName = 'RADAR-base/RADAR-Device-Auth'
+	githubUrl = 'https://github.com/' + githubRepoName + '.git'
+	issueUrl = 'https://github.com/' + githubRepoName + '/issues'
+	website = 'http://radar-base.org'
+	description = 'RADAR Device Authorizer handles authorization for data access from third party APIs for wearable devices.'
```
Comment (blootsvoets) [inline build.gradle]:
```suggestion
	description = 'RADAR REST Source Authorizer handles authorization for data access from third party APIs for for example wearable devices.'
```

File: src/main/java/org/radarbase/authorizer/config/DeviceAuthorizationConfig.java [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+ *  * limitations under the License.
+ *  *
+ *
+ */
+
+package org.radarbase.authorizer.config;
+
+import java.util.Objects;
+
+public class DeviceAuthorizationConfig {
```
Comment (blootsvoets) [inline src/main/java/org/radarbase/authorizer/config/DeviceAuthorizationConfig.java]: SourceAuthorizationConfig

File: src/main/java/org/radarbase/authorizer/config/DeviceAuthorizationConfig.java [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+    private String grantType;
+
+    private String scope;
+
+    private String clientId;
+
+    private String clientSecret;
+
+    public String getDeviceType() {
```
Comment (blootsvoets) [inline src/main/java/org/radarbase/authorizer/config/DeviceAuthorizationConfig.java]: getSourceType

File: src/main/java/org/radarbase/authorizer/config/DeviceAuthorizationConfig.java [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ */
+
+package org.radarbase.authorizer.config;
+
+import java.util.Objects;
+
+public class DeviceAuthorizationConfig {
+
+    private String deviceType;
```
Comment (blootsvoets) [inline src/main/java/org/radarbase/authorizer/config/DeviceAuthorizationConfig.java]: sourceType

File: src/main/java/org/radarbase/authorizer/config/DeviceAuthorizerApplicationProperties.java [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+
+package org.radarbase.authorizer.config;
+
+import java.util.List;
+import java.util.Objects;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.web.cors.CorsConfiguration;
+
+@ConfigurationProperties(prefix = "device-authorizer", ignoreUnknownFields = false)
```
Comment (blootsvoets) [inline src/main/java/org/radarbase/authorizer/config/DeviceAuthorizerApplicationProperties.java]:
```suggestion
@ConfigurationProperties(prefix = "radar-rest-source-authorizer", ignoreUnknownFields = false)
```

File: src/main/java/org/radarbase/authorizer/config/DeviceAuthorizerApplicationProperties.java [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+public class DeviceAuthorizerApplicationProperties {
```
Comment (blootsvoets) [inline src/main/java/org/radarbase/authorizer/config/DeviceAuthorizerApplicationProperties.java]:
```suggestion
public class RestSourceAuthorizerApplicationProperties {
```

File: src/main/resources/application.yml [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+
+  jpa:
+    database-platform: org.hibernate.dialect.PostgreSQL94Dialect
+    properties.hibernate.temp.use_jdbc_metadata_defaults: false
+
+  jackson:
+    serialization:
+      write_dates_as_timestamps: false
+
+device-authorizer:
```
Comment (blootsvoets) [inline src/main/resources/application.yml]:
```suggestion
radar-rest-source-authorizer:
```
Is this cors config still used?

File: src/main/resources/application.yml [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+  cors:
+    allowed-origins: '*'
+    allowed-methods: GET, PUT, POST, DELETE, OPTIONS
+    allowed-headers: Authorization, Content-Type
+    allow-credentials: true
+    max-age: 1800
+
+  device-clients-file-path: src/main/docker/etc/device_clients_configs.yml
```
Comment (blootsvoets) [inline src/main/resources/application.yml]:
```suggestion
  config-path: src/main/docker/etc/radar_rest_source_config.yml
```

**Pull request merged**